cmd/server/main.go

```diff
@@ -381,6 +400,17 @@ func rateLimitMiddleware(maxReq int, windowLength time.Duration) func(http.Handl
 		maxReq,
 		windowLength,
 		httprate.WithKeyFuncs(func(r *http.Request) (string, error) {
+			if session.IsAllowedAnonymous(r.Context()) {
+				// Rate limit by IP address if the user is anonymous.
+				ip := r.Header.Get("X-Real-IP")
+				if ip == "" {
+					ip = r.Header.Get("X-Forwarded-For")
+				}
+				if ip == "" {
+					ip = r.RemoteAddr
+				}
+				return ip, nil
```

Comment:
You shouldn't need to manually pull out the IP like this, [chi already does it](https://pkg.go.dev/github.com/go-chi/chi/v5@v5.0.11/middleware#RealIP). `r.RemoteAddr` should already contain what you want, as long as the chi.RealIP middleware runs first

Reply:
Gorgeous. Thank you!